src/main/scala/io/github/spark_redshift_community/spark/redshift/Utils.scala

```diff
@@ -137,7 +139,8 @@ private[redshift] object Utils {
           // Note: this only checks that there is an active rule which matches the temp directory;
           // it does not actually check that the rule will delete the files. This check is still
           // better than nothing, though, and we can always improve it later.
-          rule.getStatus == BucketLifecycleConfiguration.ENABLED && key.startsWith(rule.getPrefix)
+          rule.getStatus == BucketLifecycleConfiguration.ENABLED &&
+            (rule.getPrefix == null || key.startsWith(rule.getPrefix))
```

Comment:
I'm unable to find documentation on what it means for a rule's prefix to be null. Does that mean it applies to all directories in the bucket, or to none?  Would you mind pointing me out to some AWS docs or source code to understand this edge case?

I've also checked our twin project [spark-snowflake](https://github.com/snowflakedb/spark-snowflake/blob/master/src/main/scala/net/snowflake/spark/snowflake/Utils.scala#L140) and it seems this condition is not there.

Reply:
In fact I just setup a lifecycle policy on a whole bucket and noticed that in this case the prefix is null
![Screenshot 2019-09-18 at 09 43 32](https://user-images.githubusercontent.com/154371/65128138-29a0d100-d9f9-11e9-83d4-a2d61b7fbde8.png)

But I search the documentation and found this: https://docs.aws.amazon.com/AmazonS3/latest/API/RESTBucketPUTlifecycle.html

It says that prefix is not required and that empty prefix mean no filter so all objects of the bucket.
![Screenshot 2019-09-18 at 09 48 58](https://user-images.githubusercontent.com/154371/65128428-c6fc0500-d9f9-11e9-8ca7-6efce3fef7b4.png)

Reply:
Oh, that makes sense. I am not sure if `prefix=null` is a condition both necessary and sufficient for the Lifecycle to exist. In the sense that:
1. it seems that when a rule applies to the whole bucket then => `prefix = null` 
2. but could it be possible when some other condition is verified then => `prefix = null`? 

If 2) is the case, then `prefix = null` doesn't *necessarily* imply 1)

I've synced up with @tinaxi  and they pointed out:
1. getPrefix is Deprecated: https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/model/BucketLifecycleConfiguration.Rule.html#getPrefix-- It also mentions:

> The method returns prefix only if was was set using setPrefix(String). Otherwise, Use LifecycleFilter.
Returns the key prefix for which this rule will apply. This method should be used only if prefix was set using setPrefix(String).

so maybe we should not base Lifecycle existing on the getPrefix.

2. they found out this [PR](https://github.com/databricks/spark-redshift/pull/357/files) where the same problem was solved by upgrading the AWS package and using the `LifecycleFilter` object. Unfortunately we cannot upgrade due to https://github.com/spark-redshift-community/spark-redshift/blob/master/build.sbt#L51 

So if this is a problem of the old AWSJavaSDKVersion = 1.7.4, and the proper solution is to upgrade, then maybe the course forward is to open an issue to upgrade. We can still push this out without the one condition `rule.getPrefix == null`. The additional tests are very useful regardless.

@smoy and @tinaxi what do you think ?

Reply:
@lucagiovagnoli I agree. Let's create a backlog ticket to upgrade the AWS package. Meanwhile, I also think think this change is a beneficial.